checksum/runner_service.py

```diff
@@ -0,0 +1,268 @@
+from arteria.web.state import State as arteria_state
+import logging
+import subprocess
+import collections
+import asyncio
+
+
+log = logging.getLogger(__name__)
+
+
+class Job:
+    """
+    Class used to run a command and keep track of its status
+
+    Attributes
+    ----------
+    job_id: int
+        id of the job
+    cmd: [str]
+        command to run
+
+    Methods
+    -------
+    get_status()
+        returns current status
+    wait()
+        wait for job to complete
+    cancel()
+        cancel current job
+    """
+
+    def __init__(self, job_id, cmd, **kwargs):
+        """
+        Parameters
+        ----------
+        job_id: int
+            id of the job
+        cmd: [str]
+            command to run
+        **kwargs:
+            arguments to be forwarded to subprocess.Popen
+        """
+        self.job_id = job_id
+        self.cmd = cmd
+        self._status = arteria_state.STARTED
+        log.info(f"Starting:\n job id: {job_id}\n cmd: {cmd}")
+        log.debug(f"kwargs: {kwargs}")
+        try:
+            self._proc = subprocess.Popen(self.cmd, **kwargs)
+        except Exception as e:
+            log.error(e)
+            raise
+
+    def get_status(self):
+        """
+        Get job status.
+
+        Can be one of the following from `arteria.web.state.State`:
+            * `STARTED`
+            * `DONE`
+            * `ERROR`
+            * `CANCELLED`
+        """
+        if self._status == arteria_state.STARTED:
+            return_code = self._proc.poll()
+
+            if return_code is None:
+                self._status = arteria_state.STARTED
+            elif return_code == 0:
+                self._status = arteria_state.DONE
+                log.info(
+                    f"Job {self.job_id} completed successfully")
+            else:
+                self._status = arteria_state.ERROR
+                log.error(
+                    f"Job {self.job_id} failed with status code {return_code}")
+
+        return self._status
+
+    def wait(self):
+        """
+        Wait for the job to complete.
+        """
+        self._proc.wait()
+
+    def cancel(self):
+        """
+        Cancel the job.
+        """
+        if self.get_status() == arteria_state.STARTED:
```

Comment:
Hmm, when I wrote this function, I think I saw it as something to make sure a job is not in the running state, so from this point of view if a job is already not running when calling this function, it's all good.

But yes I see your point, how strongly do you think this should be communicated? Like with raising an exception or something more quiet like logging a message at debug level?

Reply:
Hmm good question, yeah, the main mission here is to make sure the job isn't running, so I don't really think it should be an exception. On the other hand I think it would be nice if a message about the job not being in a canceling state would be returned in the response instead of just written to the log. If a job presumed running isn't running, I would like to know about it. But I guess that the reason for that would in most cases be that the job is already done. 

However, I don't feel strongly about this, a debug message would be enough.

Reply:
I could return the current state, so if the job was in `DONE` or `ERROR` to start with we would know about it. That wouldn't work if the job was already in `CANCELLED` state though :thinking: